[Tests] Fix Android Tests Compilation Failure (`V2`)

ParaskP7 opened the pull request:
Closes: [AINFRA-1080](https://linear.app/a8c/issue/AINFRA-1080)
Related: #21627

---

## Description
<!-- Describe the changes, why they are needed, and how they address the issue -->

This PR replaces the `androidx.test.espresso.core.internal.deps.guava.base.Preconditions.checkNotNull` static import with `com.google.common.base.Preconditions.checkNotNull` to resolve the import issue.

Updating to `androidx.test:core` to  `1.7.0` (done as part of #22091) made this static `androidx.test.espresso.core.internal.deps.guava.base.Preconditions.checkNotNull` import no longer resolvable. Updating the import to `com.google.common.base.Preconditions.checkNotNull`, which comes from this `com.google:guava` transitive dependency, solve this issue.

For more info see:
- [1.7.0](https://developer.android.com/jetpack/androidx/releases/test#core-1.7.0)

> [!IMPORTANT]  
> In addition to that d720335b7793791bbf5a04b33ab60a18958848a5 commit which fixes the issue, I also added another extra 094cd91a89bcce8c26369de577f0dd0edeb7b254 commit, which makes sure that we at least run `assemble` on our `androidTest` source set and avoid these kind of issues going forward. These `wordpress` and `jetpack` assemble jobs ~~shouldn't take too much time to complete, nor too much CI resources~~, and wouldn't run the actual UI tests, so adding those shouldn't be a problem to us (🤔).


---

## Testing instructions
<!-- Consider listing specific testing steps for helping reviewers evaluate the efficacy of the changes -->

Run both the below Gradle tasks locally and make sure they complete successfully:
- `./gradlew assembleJetpackJalapenoDebugAndroidTest` (for dependency cache purposes - [related build](https://buildkite.com/automattic/wordpress-android/builds/22883))
- `./gradlew buildHealth` (for dependency analysis purposes - [related build failure](https://buildkite.com/automattic/wordpress-android/builds/22882))


---

### Post Merge Instructions

- [x] ~~Trigger [schedules/dependency-analysis.yml](https://github.com/wordpress-mobile/WordPress-Android/blob/trunk/.buildkite/schedules/dependency-analysis.yml) manually via [this](https://buildkite.com/automattic/wordpress-android) `New Build` button to run the `Dependency Analysis` job again.~~ -> [Done](https://buildkite.com/automattic/wordpress-android/builds/22891/steps/canvas)

<!--

Consider testing the following *yourself* before requesting a review:

- Compatibility with WordPress.com sites and self-hosted Jetpack sites
- Both portrait and landscape orientations
- Light and dark modes
- Dynamic font sizes: larger, smaller and bold text
- High contrast mode
- Screen reader experience (e.g., Talkback)
- Languages with large words or with letters/accents not frequently used in English
- Right-to-left languages layout
- Multi-tasking: split view and slide over

-->

<!--

Use these emoji to improve the code review

❤️ - Praise. Compliments about great solution.
⛏ - Nitpick️. Not high importance. Usually, short one-liners.
💡 - Idea. Alternative solution or improvement.
❓ - Question. Requires some explanation or more context.
⚠️ - Warning. Should be addressed in this PR or in some follow-up PR.
🛑 - Blocker. Must be addressed before PR gets merged.

-->

ParaskP7 committed:
Tests: Replace *.deps.guava.base.* with com.google.common.base.* import

Updating to 'androidx.test:core' to  '1.7.0' made this static
'androidx.test.espresso.core.internal.deps.guava.base.
Preconditions.checkNotNull' import no longer resolvable. Updating the
import to 'com.google.common.base.Preconditions.checkNotNull', which
comes from this 'com.google:guava' transitive dependency, solve this
issue.

For more info see:
- build(deps): Bump androidx.test:core from 1.6.1 to 1.7.0 #22091
https://github.com/wordpress-mobile/WordPress-Android/pull/22091
- 1.7.0: https://developer.android.com/jetpack/androidx/releases/test
#core-1.7.0

ParaskP7 committed:
CI: Add assemble android test jobs for both wordpress and jetpack

This is done because there has been now multiple occasions where a
dependency got updated, with the 'androidTest' source set failing to
compile, but since CI is no longer running instrumented tests, such
failure are not reported on the spot.

Previous Example: [Tests] Fix Android Tests Compilation Failure #21627

The main problem with this is that when this happens, at least our
weekly scheduled 'Dependency Analysis' job starts failing, and this
causes us too much trouble to get to the bottom of this. An
extra PR needs created/reviewed/tested, just to fix this small failure,
which would be otherwise fixed on the spot.

ParaskP7 commented on WordPress/src/androidTest/java/org/wordpress/android/support/WPSupportUtils.java:
```diff
@@ -65,6 +64,7 @@
 import static androidx.test.espresso.matcher.ViewMatchers.withText;
 import static androidx.test.platform.app.InstrumentationRegistry.getInstrumentation;
 import static androidx.test.runner.lifecycle.Stage.RESUMED;
+import static com.google.common.base.Preconditions.checkNotNull;
```
Thanks, but actually this is the correct order for these imports. 😊

wzieba commented on .buildkite/pipeline.yml:
```diff
@@ -153,6 +153,21 @@ steps:
+  #################
+  # Assemble Instrumented (aka UI) Tests
+  #################
+  - group: "🔬 Assemble instrumented tests"
+    steps:
+       - label: ":wordpress: 🔬 Assemble instrumented tests"
+         command: ./gradlew WordPress:assembleWordpressVanillaDebugAndroidTest
+         plugins:
+           - $CI_TOOLKIT
+
+       - label: ":jetpack: 🔬 Assemble instrumented tests"
+         command: ./gradlew WordPress:assembleJetpackVanillaDebugAndroidTest
+         plugins:
+           - $CI_TOOLKIT
```
If the product team decided to disable these tests, I'd suggest to rather exclude `androidTest` configuration from `buildHealth` and not run `assemble(Wordpress|Jetpack)VanillaDebugAndroidTest` tasks. We already have some bottlenecks on our CI queue, and running these tasks doesn't seem to have any benefit right now.

ParaskP7 replied:
Hmmm, yea, maybe you're right @wzieba . 💯

FYI: I just fear that us completely disabling any process from running anything `androidTest` related (like `buildHealth`), then after a year (or two) without them, it would be a bigger pain to have them compile and run again (if that's the plan)... 🤷

@nbradbury + @oguzkocer whyt? 🤔

nbradbury replied:
I'm not sure I have a thorough enough understanding of this, but I'm still torn on how to respond. When I hear that something will improve CI I almost immediately want to be in favor of it, but I do get @ParaskP7's point about the potential problems down the road. @oguzkocer You have a stronger background than I in infra so perhaps you can better reply?

nbradbury replied:
Just an FYI that it looks like Oguz is AFK today and tomorrow.

ParaskP7 replied:
FYI: Just to test it, I just pushed these b4ea221130e50579831cf0e67d425c237867ce9d + 4c69f53102454fcbd7ae9510c883e42b7a4b53eb commits to run those `assembleAndroidTest` tasks as part and after the prototype build. That will definitely make things faster, less resource intensive, and I just want to test how much more time it will take for the prototype build to complete.

ParaskP7 replied:
FYI: By adding `assembleAndroidTest` to prototype builds it seems that we are adding 5' extra mins to the job overall, not sure if that's acceptable (or not). 🤔

oguzkocer replied:
#21507 doesn't have much context, so I don't know why the tests were disabled in the first place.

I don't have all the context yet, but my general recommendation is to avoid making anything test related a dependency of regular builds. They should be assembled/executed in their own step, without preventing any other step. The only semi-exception to this would be for release builds where we run the lint task. In every other case, if a step is important, all we need to do is to make it a required check in GitHub PRs.

Our goal, especially now that AI is becoming a major part of our workflow, should be to have fast CI builds. I'd love to get to a state where I don't have to run a local build and instead work on multiple features with Claude and rely on the CI build artifacts to address issues. This is unfortunately not possible in WPAndroid at the moment, because it's taking forever to get the result of some of the builds. In a toy project I work on occasionally, I am even trying out the idea of leaving PR review comments to my own changes and asking Claude to read my review (using `gh` cli tool) and address the feedback. I think this kind of workflow will become more common as AI gets better and we have to do less of the coding ourselves, so I'm in favor of anything that will get me CI results faster.

nbradbury replied:
> Our goal, especially now that AI is becoming a major part of our workflow, should be to have fast CI builds

I strongly agree with this.

ParaskP7 replied:
Thanks for all the discussion folks, much appreciated! 🙇 @wzieba @nbradbury @oguzkocer 

1. I went ahead and reverted all the CI changes made: c65e8a0aff9b56f9c646941e7c265e873bdeed4d + 3bfea8f22a75d2ed34fa3be3af8ed5d0a652c55d + c65e8a0aff9b56f9c646941e7c265e873bdeed4d
2. And as per @wzieba suggest I added a configuration to ignore all `test` related analysis, including unit tests, more details on the commit description itself: b57fefaa2a7745bc05b484c22a59a767017c46ea

Let me know if I can now proceed merging this. 🙏

ParaskP7 committed:
CI: Run assemble android test as part of and after a prototype build

Running the 'assembleAndroidTest' task as part of a separate job is too
slow and resource intensive. Running the 'assembleAndroidTest' task as
part of and after a "Prototype Build", which would alread have already
run the 'assemble' task before and thus have everything set might make
the whole process much faster.

ParaskP7 committed:
CI: Replace vanilla with jalapeno for assemble android test on prototype

This is because the prototype build are not using the 'Vanilla' product
flavor, but rather the 'Jalapeno', see:

PROTOTYPE_BUILD_FLAVOR = 'Jalapeno'
PROTOTYPE_BUILD_TYPE = 'Debug'

FYI: This should make it much faster as there will not be the need to
assemble another product flavor, from scratch, and before running
assemble on Android tests.

ParaskP7 committed:
Revert "CI: Replace vanilla with jalapeno for assemble android test on prototype"

This reverts commit 4c69f53102454fcbd7ae9510c883e42b7a4b53eb.

ParaskP7 committed:
Revert "CI: Run assemble android test as part of and after a prototype build"

This reverts commit b4ea221130e50579831cf0e67d425c237867ce9d.

ParaskP7 committed:
Revert "CI: Add assemble android test jobs for both wordpress and jetpack"

This reverts commit 094cd91a89bcce8c26369de577f0dd0edeb7b254.

ParaskP7 committed:
Build: Ignore test analysis for dependency analysis plugin

PR Comment: https://github.com/wordpress-mobile/WordPress-Android/pull/
22096#discussion_r2251439997

FYI: This is done because the 'assembleAndroidTest' CI jobs are being
indefinitely commented out and causing 'buildHealth' to fail compilation
now and then. By ignoring tests altogether we avoid this problem, at
least temporarily, and until we find a better way to deal with this
problem, or else, get the 'assembleAndroidTest' CI jobs run again.

PS: I couldn't just ignore 'androidTest' and had to ignore all 'test'
from analysis because there isn't any way to ignore only 'androidTest'
specifically. I also tried the below 'ignoreSourceSet('androidTest')'
approach, but this too didn't work as expected because it tried to
compile the 'androidTest' source set anyway, failing the build on errors
like this 'import' error we fixed before.

dependencyAnalysis {
    issues {
        ignoreSourceSet('androidTest')
    }
}

Unfortunately, the above configuration is documented only for JVM
sources and not directly for Android-specific source sets like
androidTest. Recent issue discussions and feature requests have
requested more granularity, such as being able to configure this for any
source set, but as of the latest public information, this remains a
limitation.

Discussion: Update dependencyAnalysis configuration to allow for
sourceset selection for all projects #901
-  https://github.com/autonomousapps/dependency-analysis-gradle-plugin/
issues/901

wzieba reviewed (approved):
Thanks for providing details in the commit message. I think this is a good decision! 👍
